blocks/graph/bn.py

```diff
@@ -0,0 +1,163 @@
+        of accumulated population statistics.
+    update_pairs : list of tuples
+        A list of 2-tuples where the first element of each tuple is the
+        shared variable containing a "population" mean or standard
+        deviation, and the second is a Theano variable for the
+        corresponding statistics on a minibatch. Note that multiple
+        applications of a single :class:`blocks.bricks.BatchNormalization`
+        may appear in the graph, and therefore a single population variable
+        may map to several different minibatch variables.
+
+    See Also
+    --------
+    :func:`batch_normalization`, for an alternative method to produce
+    batch normalized graphs.
+
+    """
+    # Avoid circular imports.
+    from blocks.bricks import BatchNormalization
+    from ..filter import VariableFilter, get_application_call
+
+    # Create filters for variables involved in a batch normalization brick
+    # application.
+    def make_variable_filter(role):
+        return VariableFilter(bricks=[BatchNormalization], roles=[role])
+
+    # Group inputs and outputs into dicts indexed by application call.
+    def get_app_call_dict(variable_filter):
+        return collections.OrderedDict((get_application_call(v), v) for v in
+                                       variable_filter(computation_graph))
+
+    # Compose these two so that we get 4 dicts, grouped by application
+    # call, of different variable roles involved in BatchNormalization.
+    inputs, outputs, means, stdevs = map(get_app_call_dict,
+                                         map(make_variable_filter,
+                                             [INPUT, OUTPUT, BATCH_NORM_OFFSET,
+                                              BATCH_NORM_DIVISOR]))
+
+    assert len(set([len(inputs), len(outputs), len(means), len(stdevs)])) == 1
+
+    # Remove any ApplicationCalls that were not generated by apply(), or
+    # were generated by an apply() while already in training mode.
+    remove = filter(lambda a: (a.metadata.get('training_mode', False) or
+                               a.application.application !=
+                               BatchNormalization.apply), inputs.keys())
+    for app_call in remove:
+        for mapping in (inputs, outputs, means, stdevs):
+            del mapping[app_call]
+
+    replacements = []
+    update_pairs = []
+    for app_call in inputs:
+        old_output = outputs[app_call]
+        unpacked = inputs[app_call].owner.inputs[0]
+        with app_call.application.brick:
+            new_output = app_call.application.brick.apply(unpacked)
+        replacements.append((old_output, new_output))
+        new_app_call = get_application_call(new_output)
+        update_pairs.append((app_call.application.brick.population_mean,
+                             new_app_call.metadata['offset']))
+        update_pairs.append((app_call.application.brick.population_stdev,
+                             new_app_call.metadata['divisor']))
```

Comment:
In the BatchNormalization constructor, these metadata are set to the population statistics, and so the updates don't change anything. Am I missing something?